### PR TITLE
6919 preview tools

### LINKED ISSUE #6919: Preview Tools - New external tool preview type config setting, icon btn on dataset/file pgs
**PROBLEM**

_How do I set a preview tool ONLY, with no explore tool btn displayed?_

When a sysadmin configures one of the Dataverse Previewers (e.g. Text Preview or PDF Preview) external tools to their installation, it adds not only a "Preview" tab with embedded tool view to the file pg, but also an "Explore" button to both the file pg and the dataset pg (the same Explore btn tools like Two Ravens, World Map and Data Explorer are linked from, see screenshot below, [example on Harvard Dataverse](https://dataverse.harvard.edu/file.xhtml?persistentId=doi:10.7910/DVN/A0HHDI/CUTVRJ&version=11.0)).

In our UI, both the preview and explore workflows are linked to the same tool. This implies the same tool provides two features, when in fact all the Dataverse Previewers external tools configured in Harvard Dataverse provide only one "preview" feature. When you click the "Explore" btn, you are provided the exact same view of the file that is embedded on the file pg, with no additional features, with the exception of the citation and download btn -- the same information also provided on said file pg. The full version of the Dataverse Previewers external tools look so similar to the file pg, in fact, that I question the value it provides the user when they click the "Explore" btn.

Screen Shot 2020-05-19 at 4 05 07 PM

**PROPOSAL**

I suggest there is more value having users go to the file pg as opposed to clicking Explore and go to the landing pg of an external preview tool when said tool is embedded directly on the file pg. As such, we should remove preview tools from being linked to the Explore btn of the dataset and file pgs. In the Preview tab, the "Explore on {PREVIEW TOOL NAME}" btn should instead be labeled "Open preview tool in new window..." or something more succinct, to remove any implication of additional features when clicking the button.

### PR DESCRIPTION
**Pull request by pdurbin**
**What this PR does / why we need it**:

This pull request does two things:

- The Preview tab is available in more cases, such as...
  - when a guestbook is present and
  - when terms are present.
- External tools can now be "preview only" in the sense that...
  - when preview only tools appear on the Preview tab (on the file page), a button says "Open in New Window" rather than "Explore on [TOOL NAME]" and
  - preview only tools do not appear under "Explore Options" under the "Access File" dropdown button.

Prior to this pull request, the Preview tab was only available for files that are publicly downloadable.

The way to create a "preview only" tool is to use the new "preview" type.

~~Note that explore tools can continue to appear under the Preview tab (on the file page) if they have hasPreviewMode set to true, as before.~~ This was removed in 767882b

**Which issue(s) this PR closes**:

Closes #6919 (Preview Tools - New external tool preview type config setting, icon btn on dataset/file pgs)

Some commits have been added to address #4429 (Revise Explore Tools documentation to reflect modularity )

**Special notes for your reviewer**:

Stuff to notice:

- Custom UIInput fields have been removed.
- We've switched to standard Bean Validation.

Some questions I have:

- Should we add email validation to the guestbook form?

**Suggestions on how to test this**:

The code having to do with the terms/guestbook popup was extensively reworked so it would be good to test...

- validation of the form
- that the values entered actually go into the database (don't laugh)
- all the various places that the popup can appear
  - dataset page
  - file page
  - when clicking external tools
  - when clicking package files
  - when clicking worldmap files

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

## Screenshot 1: preview only example

Below is an example of a "preview only" tool. Note that the button in the Preview tab says "Open in New Window" and the tool is not present under the "Access File" dropdown.

Screen Shot 2020-10-06 at 1 49 03 PM

## Screenshot 2: Terms + Guestbook

Below is an example of terms and a guestbook being presented in a Preview tab. The tool happens to be an explore tool and shows up under "Explore Options" under the "Access File" dropdown.

Screen Shot 2020-10-06 at 1 48 07 PM

**Is there a release notes update needed for this change?**:

Yes, a release note has been included.

**Additional documentation**:

Doc changes included.

### CODE REVIEW

File: doc/sphinx-guides/source/api/external-tools.rst [hunk 1]
```diff
@@ -70,19 +78,19 @@ Terminology
...
 
     toolUrl                      The **base URL** of the tool before query parameters are added.
     
-    hasPreviewMode               A boolean that indicates whether tool has a preview mode which can be embedded in the File Page. Since this view is designed for embedding within Dataverse, the preview mode for a tool will typically be a view without headers or other options that may be included with a tool that is designed to be launched in a new window. Sometimes, a tool will exist solely to preview files in Dataverse and the preview mode will be the same as the regular view. 
+    hasPreviewMode               A boolean that indicates whether tool has a preview mode which can be embedded in the File Page. Since this view is designed for embedding within Dataverse, the preview mode for a tool will typically be a view without headers or other options that may be included with a tool that is designed to be launched in a new window.
```
Comment (sekmiller) [inline doc/sphinx-guides/source/api/external-tools.rst:94]: Do we want/need to say that hasPreviewMode only pertains to explore tools?

Comment (pdurbin) [reply doc/sphinx-guides/source/api/external-tools.rst:94]: As we discussed with @scolapasta we might do away with hasPreviewMode entirely and factor out "type" to its own table. The idea would be the perhaps tools could have multiple types, so an explore tool would also be a preview tool if it supports preview. We plan to discuss more next week.

Comment (qqmyers) [reply doc/sphinx-guides/source/api/external-tools.rst:94]: FWIW: Looking at 5.x with QDR, some of the previewers like audio are small (regardless of file size) and it really makes no sense to have a separate explore button on the file page, so being able to mark such a previewer as preview only would help there, and would make it easy for admins to decide whether other previewers are worth popping out as explore tools or not. So - support for allowing tools to register with one or more types.

Comment (pdurbin) [reply doc/sphinx-guides/source/api/external-tools.rst:94]: @qqmyers thanks. To be clear, even with the code as-is in this pull request, it's possible for admins to decide if a tool that offers preview appears under "explore" or not. To convert the audio tool to "preview only" (again, given the code so far), you'd change it from an "explore" tool to a "preview" tool. To switch it back, you'd change it from "preview" to "explore" and make sure hasPreviewMode is true.

Also, please note that from the Preview tab, there's always a button to open the tool in a new window. For preview tools the button is labeled "Open in a New Window" and for explore tools it's labeled as it was before, "Explore on [TOOL NAME]". So popping out happens for both.

**Comment by TaniaSchlatter** [comment]
After review, let's please take out the word "Tool" from the Preview button.

**Comment by pdurbin** [comment]
@TaniaSchlatter can do. Here's a screenshot of the "before":

Screen Shot 2020-10-15 at 10 43 26 AM

**Comment by pdurbin** [comment]
As I mentioned during sprint planning yesterday, we'd like to do the following additional development before more code review and QA:

- ~~Remove "Tools" from preview tool switcher.~~ Done in e5eacae
- Write SQL migration script for external tools.
  - Drop "hasPreviewMode" from externaltool table.
  - Move "type" from externaltool table to its own table, allowing multiple types.
  - Tools with hasPreviewMode=true should be given a type of "preview" (and "explore"?).
- Change manifest parsing to allow "type" to accept an array (new way) as well as a string (old way) for backward compatibility.
- Refactor all code to look for the "preview" type only and not hasPreviewMode.
- Update docs to reflect multiple types.

(If I'm forgetting anything, please let me know!)

The other thing that's on my mind is that I left a comment in the guestbook code saying that we could do email address validation on the email field (now that we've switched to Bean Validation) but I'm going to consider this out of scope.

**Comment by pdurbin** [comment]
In 767882b I implemented multiple types and removed hasPreviewMode. I'm sending this back into code review.

File: doc/sphinx-guides/source/api/external-tools.rst (hunk 1 above)
Comment (pdurbin) [reply doc/sphinx-guides/source/api/external-tools.rst]: hasPreviewMode has been removed in 767882b

File: src/main/webapp/file-download-popup-fragment.xhtml [hunk 2]
```diff
@@ -121,56 +123,83 @@
...
+                <f:param name="DO_GB_VALIDATION_#{popupContext}" value="true"/>
             </p:commandButton>
+            <!--EXTERNAL TOOL BUTTON-->
             <!--On the dataset page (but not the file page), "tool" is null so we get the tool from the guestbookResponse.-->
-            <p:commandButton styleClass="btn btn-default" process="@this" value="#{bundle.acceptTerms}" rendered="#{guestbookResponse.fileFormat == 'externalTool'}"
-                           action="#{fileDownloadHelper.writeGuestbookAndLaunchExploreTool(guestbookResponse, fileMetadata, tool)}"
-                           update="guestbookUIFragment">
-                <f:param name="DO_GB_VALIDATION" value="true"/>
+            <p:commandButton styleClass="btn btn-default" value="#{bundle.acceptTerms}" rendered="#{guestbookResponse.fileFormat == 'externalTool'}"
+                             action="#{fileDownloadHelper.writeGuestbookAndLaunchExploreTool(guestbookResponse, fileMetadata, tool)}"
```
Comment (sekmiller) [inline src/main/webapp/file-download-popup-fragment.xhtml:191]: If we are rendering the buttons based on the gbResponse file formats, can we just have one method for Launching the explore tool, world map and package popup that decides what to do based on the file format. That could reduce three buttons to one.

**Review by sekmiller — COMMENTED** [review]
Looks good overall. I just have some suggestions for code cleanup/consolidation.

File: src/main/webapp/file-download-popup-fragment.xhtml (hunk 2 above)
Comment (scolapasta) [reply src/main/webapp/file-download-popup-fragment.xhtml:191]: A related thought: @djbrooke  has given the ok to start removing world map related buttons and such, so if that helps simplify the code... (I've started to do that in the kebab issue / edit buttons)

Comment (djbrooke) [reply src/main/webapp/file-download-popup-fragment.xhtml:191]: Yes - I have on my list to start clearing out Worldmap issues from Github and to create an issue to pull out any Worldmap/Geoconnect code specific in Dataverse, so it's good to get a headstart.

File: src/main/java/edu/harvard/iq/dataverse/FileDownloadHelper.java [hunk 3]
```diff
@@ -230,6 +133,7 @@ public void writeGuestbookAndStartDownload(GuestbookResponse guestbookResponse)
          boolean valid = validateGuestbookResponse(guestbookResponse);
 
          if (!valid) {
+             // FIXME: This never validation error shows in the UI (even if you add a bundle entry).
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/FileDownloadHelper.java]: Done in e73d2f4.

File: src/main/java/edu/harvard/iq/dataverse/FileDownloadHelper.java [hunk 4]
```diff
@@ -347,7 +251,22 @@ public String startWorldMapDownloadLink(GuestbookResponse guestbookResponse, Fil
...
     }
 
-        
+     /**
+      * Writes a guestbook entry for either popup scenario: guestbook or terms.
+      */
+     public boolean writeGuestbookAndShowPreview(GuestbookResponse guestbookResponse) {
+         // We've already validated guestbooks on the front end but we validate again
+         // on the backend (that's what older methods do).
+         boolean valid = validateGuestbookResponse(guestbookResponse);
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/FileDownloadHelper.java]: Done in e73d2f4.

File: src/main/java/edu/harvard/iq/dataverse/FileDownloadHelper.java [hunk 5]
```diff
@@ -230,6 +133,7 @@ public void writeGuestbookAndStartDownload(GuestbookResponse guestbookResponse)
          boolean valid = validateGuestbookResponse(guestbookResponse);
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/FileDownloadHelper.java]: Done in e73d2f4.

File: src/main/webapp/file-download-popup-fragment.xhtml (hunk 2 above)
Comment (pdurbin) [reply src/main/webapp/file-download-popup-fragment.xhtml:191]: I just looked at this and I don't see a lot of value in refactoring. The complexity has to live somewhere and right now a decision is made on the front end. I think it's fine as it is.

**Comment by pdurbin** [comment]
I addressed code review comments in e73d2f4 and I'm moving this back to code review.

**Review by sekmiller — APPROVED** [review]

**Comment by kcondon** [comment]
@pdurbin I accidentally merged, then reverted this pr. In the process of attempting to deploy, I'd found that it appears to conflict with the mail regex branch on flyway script names. Mail regex is ready to be merged and preview still needs to be tested so was hoping you could fix the script name here. I don't think we can reopen this pr but the preview branch still exists. Can you fix there and make a new pr?

**Comment by pdurbin** [comment]
@kcondon yes, I just created pull request #7369 which replaces this one.

**Pull request merged**